### DISCUSSION TITLE
Polling and storing feature flags in cache

### DISCUSSION THREAD

**Discussion by IrinaSouth** [featurehub-io/featurehub#1055] [Q&A] [answered]
I'd like to store all features in cache. I need to refresh this cache on a daily basis, specifically just once a day. How may I poll all of my features at once?

**Comment by rvowles** [comment]
There are two ways to do this, the easiest way to do it is to use a pull-cache - so you have some CI task or similar that goes and grabs the features and saves them in some location - like creating a dynamic PR/MR which will trigger your downstream pipelines. The alternative way is to use the FeatureHub WebHooks, they give you the same essential data but you need to have a cloud function or lamba or equivalent running and then manipulate the data.


So lets assume that you have just created a new instance of FeatureHub, 

````
docker run --rm -p 8085:8085 featurehub/party-server:latest
````

You create the initial user and portfolio, create the first application and then start adding features:

![image](https://github.com/featurehub-io/featurehub/assets/55768/323b46ef-c31b-404c-a199-73f07a3d370c)

and so say now we have several features (two in this case) 
![image](https://github.com/featurehub-io/featurehub/assets/55768/b33c2cd2-299e-4ce2-adf9-c027bc17678b)

you then go and create a service account that is allowed to read these features in this environment (production)

![image](https://github.com/featurehub-io/featurehub/assets/55768/983a2e77-8f07-4c42-bcb1-250a01f9c259)

then you allow it the access

![image](https://github.com/featurehub-io/featurehub/assets/55768/a834d973-9c0c-44e6-95f3-20439bc895c0)

![image](https://github.com/featurehub-io/featurehub/assets/55768/1170d4ec-65d1-4848-a8f7-93c95a843332)

and now you can go to the API Keys section and grab the key - if you will need to grab the client key as it will have all of the rules in it. 

![image](https://github.com/featurehub-io/featurehub/assets/55768/9113c118-108c-4226-a622-0629ccbf69db)

Copy that Client SDK key and then you can type in something like

```
curl -v 'localhost:8085/features?apiKey=f1383cf5-2128-48f6-84af-e93da30ae6ba/MqAiuaWaryFYDULE8kR7v2J1jj1xxR*hpPF0V6rTxdRgZ2r0w8Z&contextSha=0'
```

That long set of characters is the API key. This will give you something like:

````
[{"id":"f1383cf5-2128-48f6-84af-e93da30ae6ba","features":[{"id":"41221e1c-73d0-4717-82d2-610068f8c4fe","key":"project123_ticket_2","l":true,"version":1,"type":"BOOLEAN","value":false,"strategies":[]},{"id":"cdc25cbb-f538-4768-8aca-26e4cf5f172c","key":"proj123_some_ticket","l":true,"version":1,"type":"BOOLEAN","value":false,"strategies":[]}]}]
````

This you can save whereever you need to cache it.

When you want to load it into your SDK, give it the same key (so it knows it is using a client-sdk key) but then get the repository (typically something like config.repository()).

Decode this string into a JSON set of data (depending on the SDK you are using there may be some help in this, JS/C#/Java/Go, Python and Ruby don't) and get the the "features" array, thats the actual JSON data you need to pass to the repository for it to start functioning. The SDK doesn't need to poll in the SDK itself, all of the SDKs have the basics required to just interpret the data in the repository per user.

**Reply by Avivma** [reply]
Thank you for the explanation.
Unfortunately, it doesn't work for me. I get 404.
I created a new user and portfolio. New application and features. I created it using FeatureHub server (I didn't create new instance using docker).
The url I'm using is:
https://app.featurehub.io/features?apiKey=my-api-key
Should this work?... If not what host should I use?

(replaced my real api key with "my-api-key". In my request I paste the real one - copied from the client api key)

**Reply by IrinaSouth** [reply]
Hi there, if you are using our SaaS platform, then the Base URL should be what you see on the API Keys page as below: 


<img width="740" alt="Screenshot 2023-09-08 at 09 26 37" src="https://github.com/featurehub-io/featurehub/assets/9538697/0e62e0cd-d75b-4b9b-8b0a-1b5674ae7a48">

**Reply by Avivma** [reply]
Good morning,
I tried this url:
https://app.featurehub.io/vanilla/features?apiKey=my-api-key
It doesn't work also (got error code 400).
What do I missing?
host = SDK base url
path = features
param = apikey=my-api-key (I used the real one in my request)

Is there a way to get _features_ by using simple Get request (when using the your SaaS platform)?

**Reply by rvowles** [reply]
hiya - you need to add &contextSha=0 at the end. so:

curl https://app.featurehub.io/vanilla/features?apiKey=my-api-key&contextSha=0

and that is a simple GET request.

**Reply by Avivma** [reply]
Thank you!
That works.
Does it possible to get the same behavior / results with the Python SDK?
(I review the Python sdk, but I haven't seen a way to get this results. Client_context has several api, but as far as I understood it, it doesn't have this kind of api).

Also, may you please give me an example for a simple Get request for one feature with a target rule (using curl. I know this result can be done with the Python sdk).

**Comment by rvowles** [comment]
Hiya - the Python SDK has a PollingEdgeService class - you can see the code here where it does the call out? 

https://github.com/featurehub-io/featurehub-python-sdk/blob/main/featurehub_sdk/polling_edge_service.py#L73

given a successful response, you can see it just decodes it and passes it to a method:

https://github.com/featurehub-io/featurehub-python-sdk/blob/main/featurehub_sdk/polling_edge_service.py#L83

which parses it and then passes it to the repository.

https://github.com/featurehub-io/featurehub-python-sdk/blob/main/featurehub_sdk/polling_edge_service.py#L156

If you do the same call and save the resulting json instead and when you load it, mimic the method and then load it into the repository using the same mechanism that will do what you want I think?

**Comment by Avivma** [comment]
Thank you for the response!
I have more following questions:

1. If I understand your response correctly, you mean to mimic the mechanism from your sdk to my own, right?
Which mean, there isn't such a python api out-of-the-box, right?

2. How may I get the result of features with rules?
Meaning: take into account the rules' value.
Example: lets say, I have 2 features with different rules:
Feature **A**, default=False, rule **color==blue -> True**
Feature **B**, default=False, rule **fruit==apple -> True**
Does it possible to get the result of these 2 feature, while taking into account there rules, in one request?
If possible, may you give me an example?

**Comment by rvowles** [comment, marked as answer]
1) Yes, there isn't anything from  the Python SDK out of the box to save the response to somewhere, no. It should only be a few lines of Python however, which is a good thing!

2) I'm going to have to give you some adhoc code here rather than something specifically out of the box working, so here goes:

````
# provide a proper edge url and client feature key, the client key is particularly important

fh_config = FeatureHubConfig(edge_url, [client_eval_key])

# now do the code in _process_successful_results, i.e. read in your features and pass then to

fh_config.repository().notify("features", <data>)

# now you can use the config as normal, it has the features, it knows that it is client evaluated and won't try
# and go back to the server for data

ctx = fh_config.new_context().attribute_values("color", ["blue']).attribute_values("fruit", ["apple"]).build()

if ctx.feature('FEATURE_A').get_flag:
...
````

**Comment by Avivma** [comment]
Thank you very much :)
I will check it. It is very helpful
